### ISSUE TITLE
0.36 CRA-Demo Script and Feedback

### ISSUE THREAD

**Issue by micahgodbolt** [microsoft/FluidFramework#5478]
# @fluid-example/cra-demo 0.36 release

get here: [aka.ms/fluid-cra](https://aka.ms/fluid-cra)
[goto demo](#issuecomment-796233650) 
## Review of feedback from [0.35](https://github.com/microsoft/FluidFramework/issues/5253) demo

**Highlights**
> I like the approach. It'll be familiar and comfortable to React developers

> "Lastly, open up the App.tsx file, as that will be the only file we need to edit." Okay, that makes me very happy. :)

>  I think this is a great step towards making Fluid much easier to use and integrate in React apps.

**Todos**
- [x] Add overview of instructions at top
- [x] Missing yarn instructions
- [x] Conflicting ports between CRA and Tinylicious (fixed in main)
- [x] We don't explain what a container or DataObject is
- [x] "100-200 level user should not have to know about factories."
- [x] "Too many new concepts, registryEntry, Factory, Factory.type"
- [x] Jumping around in the code is hard to follow (re-order steps to follow code flow)
- [x] React hooks add too much complexity for 100/200 level course
- [x] Simplify typings for brevity
- [ ] Explain usecase for KVpair (its more of a demo) and what to do when you need something more complex
- [ ] "why on earth do we have to define the same data structure twice"
- [ ] "Why two imports? Can it be just one?"
- [ ] "An API that requires as much boilerplate code as this one just to use a single shared time property is not adoption-ready."
- [ ] Create/get flow is complicated

<details>
<summary><b>Version 0.36 Change Log</b></summary>

- #5218 Use service objects in getContainer and FluidStatic 
  - This opens the door for using services other than tinylicious
- #5414 Rename remaining usages of document to container in FluidStatic 
```tsx
Fluid.createDocument  
Fluid.getDocument

// is now

Fluid.createContainer
Fluid.getContainer
```
- #5424 Pass data object classes rather than factories to FluidStatic methods
```tsx
Fluid.createContainer(containerId, [KeyValueInstantiationFactory.registryEntry])
// is now
Fluid.createContainer(service, containerId, [KeyValueDataObject]) 

fluidDocument.createDataObject(KeyValueInstantiationFactory.type, 'kvpairId')
// is now
fluidDocument.createDataObject(KeyValueDataObject, 'kvpairId')
```
- #5474 update CRA demo to reflect 0.36 updates and other feedback
  - Add yarn instructions
  - Add TinyliciousService to import and container creation
  - update getContainerId explanation
  - Re-order steps to be more top to bottom
  - Remove hook concept to reduce complexity
  - Simplify useEffect section to reduce number of steps
  - Add links to Fluid concepts
  - Note fix to tinylicious port coming in 0.37

</details>

## Instructions

- Follow steps below
- Leave feedback below

\* This demo is a work in progress. There will be rough sections that need refactoring or refinement

**Comment by micahgodbolt** [comment]
Actual script still in PR #5474

**Comment by waystilos** [comment]
Getting this error when I run the demo:

Unhandled Rejection (Error): Request failed with status code 400

```
  if (retryAfterSeconds !== undefined && canRetry) {
        return new ThrottlingError(errorMessage, retryAfterSeconds, statusCode);
     }
    return new GenericNetworkError(errorMessage, canRetry, statusCode);
  }
```

**Comment by micahgodbolt** [comment]
@waystilos what part of the demo were you getting that error? Also, are you running a recent version of node? I believe Fluid requires 12+

**Comment by waystilos** [comment]
> @waystilos what part of the demo were you getting that error? Also, are you running a recent version of node? I believe Fluid requires 12+

Using node version 14.16.0

After starting tinylicious and start up the react app, it throws that error on the browser.

**Comment by skylerjokiel** [comment]
@waystilos, as @micahgodbolt said when your run the Tinylicious server you need node 12+. This is due to a security issue with earlier versions of node.

**Comment by micahgodbolt** [comment]
@waystilos Looks like the problem is starting the application too soon. This demo won't compile until the code is complete and tinylicious is running. That's a change we're hoping to make for next demo.

**Comment by Vivihung** [comment]
Thanks for sharing the steps. 
For step 4 and 5, I have a suggestion.

React has multiple ways to manage state. It depends on the developer to design their React state - shared state (redux or context) or component state. Having examples for each case will be very helpful to the community. It can guide React developers to integrate Fluid data object with React state they choose.

Above example shows how to leverage component state with useEffect() to handle the async call for Fluid data object.
We could have another example to use React context for scenarios that need to share states across components.

Something like this (I code in this textbox... there might be syntax error or typo...)
**fluidContext.tsx**
```TypeScript
import * as React from 'react';
import { createContext, useEffect, useState } from 'react';
import { MyFluidDataObject } from '../data-objects/MyFluidDataObject ';
import { getContainerId } from '../utils';
import { Fluid } from '@fluid-experimental/fluid-static';
import { TinyliciousService } from '@fluid-experimental/get-container';

export const fluidAppDataObjectId = 'kvpairId';

interface IFluidContextProviderProps {
    children: React.ReactNode;
}

type FluidContextState = {
    containerId: string;
    appData: MyFluidDataObject | null;
};

const FluidContext = createContext<FluidContextState>({
    containerId: '',
    appData: null,
});

const FluidContextProvider = ({ children }: IFluidContextProviderProps) => {
    const [state, setState] = useState<FluidContextState>({
        containerId: '',
        appData: null,
    });

    useEffect(() => {
        const { containerId, isNew } = getContainerId();
        const service = new TinyliciousService()

        async function start(): Promise<void> {
            const fluidContainer = isNew
                ? await Fluid.createContainer(service, containerId, [MyFluidDataObject])
                : await Fluid.getContainer(service, containerId, [MyFluidDataObject]);

            const appData: MyFluidDataObject = isNew
                ? await fluidContainer.createDataObject(
                    MyFluidDataObject,
                      fluidAppDataObjectId
                  )
                : await fluidContainer.getDataObject(fluidAppDataObjectId);

            setState({ containerId, appData });
        }

        start().catch((error) => console.error(error));
    }, []);

    return <FluidContext.Provider value={state}>{children}</FluidContext.Provider>;
};

function useFluidContext() {
    const context = React.useContext(FluidContext);

    if (context === undefined) {
        throw new Error('FluidContext not found');
    }

    return context;
}

export { FluidContextProvider, useFluidContext, FluidContext };
```

**App.tsx**
```TypeScript
const App = () => {
    
    return (
        <div className="App">
                <FluidContextProvider>
                    <SyncTimer />
                </FluidContextProvider>
        </div>
    );
};

export default App;
```

**SyncTimer.tsx**
```TypeScript
const SyncTimer = () => {
    const { containerId, appData } = React.useContext(FluidContext);
    const [data, setData] = React.useState<{ [key: string]: any }>({});

    React.useEffect(() => {
            const updateData = () => setData(appData.query());
            updateData();
            appData.on("changed", updateData);
            return () => { appData .off("change", updateData) }
    }, [appData]);
    return (
        <>
            <button onClick={() => appData.set("time", Date.now().toString())}>
                click
            </button>
            <span>{data["time"]}</span>
        </>
    );
};

export default SyncTimer;
```

**Comment by skylerjokiel** [comment]
@Vivihung thanks for the feedback! We've explored using Context in the past. We will definitely be using React Context patterns in our more complicated examples/tutorial.

On a related note, the fact that you took our example and modified it to use React Context makes me very happy. 😄 Our primary focus is providing an adaptable Data layer that integrates well with view libraries like React. We will be relying on area experts like yourself to help us define the best patterns!

**Comment by reinvanleirsberghe** [comment]
Why the use of the isNew variable?
In my opinion, this is the job of the fluidPackage to know that and the container should be created on the fly when needed.
If I do 
`
const fluidContainer = await Fluid.getContainer(service, containerId, [MyFluidDataObject])
`
and the containerId does not exists, the Fluid package should create one for me.
Same goes for the dataObject.

At the moment I need to set the isNew on true first to create one, and afterwards I need to set it to false, so I can use it after refresh. Or am I missing something here?

**Comment by astegmaier** [comment]
Thanks for putting this together - I'm new to the fluid framework, and this was useful to help me start wrapping my head around it.

One problem I found that it is broken with the latest versions of the `@fluid-experimental` packages (currently `0.37.4`). I tried fixing the version of each package to `0.36.1`, and it worked.

It looks like the new packages have a different import style for `Fluid` (e.g. `import Fluid from "@fluid-experimental/fluid-static";` instead of `import { Fluid } from "@fluid-experimental/fluid-static";`), and the signature of `createContainer`/`getContainer` has changed to only accept two parameters instead of three.

One other stumbling block was that it only works if you start the `tinylicious` server first so that it claims port 3000, and then CRA is smart enough to switch to port 3001. If you do it the other way around (start CRA first), tinylicious blows up. It might be worth thinking about this explicitly in a future version of this tutorial.

**Comment by skylerjokiel** [comment]
Thanks @astegmaier . You are correct 0.37 introduced some breaking changes. I'm closing this issue for now since it's no longer functional and we will work to publish the next version.

We've also addressed the tinylicious issue in version 0.37.

**Comment by micahgodbolt** [comment]
@astegmaier thanks for the feedback. I've ran into the tinylicious order issue too, though we've resolved that in current releases where the ports don't overlap any longer. Working right now on an upgraded demo that'll reflect those new changes

**Comment by micahgodbolt** [comment]
@reinvanleirsberghe you aren't missing anything....this "isNew" approach is the easiest way we've found to simulate the creation of new documents without adding a bunch of boilerplate UI to the demo. Fluid is all about co-authoring a specific "document" and in most applications there will be a very specific 'creation' UI vs 'file open' UI (or new game, join game). We totally understanding this dev experience is a bit annoying for a demo of this size, and we're always discussing ways to improve it without hiding the fact that we need to somehow create a document.

**Comment by micahgodbolt** [comment]
Closing this issue as we're moving on with 0.38 https://github.com/microsoft/FluidFramework/issues/5904
